File: Harvest/Views/Components/ValuesRadarCard.swift
import SwiftUI

struct ValuesRadarCard: View {
    let title: String
    let subtitle: String?
    let primary: AxisScores
    let primaryLabel: String
    let primaryColor: Color
    let secondary: AxisScores?
    let secondaryLabel: String?
    let secondaryColor: Color
    let onEmptyTap: (() -> Void)?

    init(
        title: String = "Your Values Map",
        subtitle: String? = nil,
        primary: AxisScores,
        primaryLabel: String,
        primaryColor: Color = HarvestTheme.Colors.primary,
        secondary: AxisScores? = nil,
        secondaryLabel: String? = nil,
        secondaryColor: Color = HarvestTheme.Colors.accent,
        onEmptyTap: (() -> Void)? = nil
    ) {
        self.title = title
        self.subtitle = subtitle
        self.primary = primary
        self.primaryLabel = primaryLabel
        self.primaryColor = primaryColor
        self.secondary = secondary
        self.secondaryLabel = secondaryLabel
        self.secondaryColor = secondaryColor
        self.onEmptyTap = onEmptyTap
    }

    private let axes: [ValueAxis] = [
        .emotionalIntelligence,
        .stability,
        .integrity,
        .connection,
        .growth
    ]

    var body: some View {
        GlassCard {
            VStack(alignment: .leading, spacing: HarvestTheme.Spacing.md) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(HarvestTheme.Typography.h4)
                        .foregroundStyle(HarvestTheme.Colors.textPrimary)
                    if let subtitle {
                        Text(subtitle)
                            .font(HarvestTheme.Typography.bodySmall)
                            .foregroundStyle(HarvestTheme.Colors.textSecondary)
                    }
                }

                if primary.isZero && (secondary?.isZero ?? true) {
                    emptyState
                } else {
                    chart
                    legend
                }
            }
        }
    }

    // MARK: - Subviews

    private var emptyState: some View {
        VStack(spacing: HarvestTheme.Spacing.sm) {
            Image(systemName: "chart.dots.scatter")
                .font(.system(size: 32))
                .foregroundStyle(HarvestTheme.Colors.accent)
            Text("Answer a few questions to map your values.")
                .font(HarvestTheme.Typography.bodyRegular)
                .multilineTextAlignment(.center)
                .foregroundStyle(HarvestTheme.Colors.textSecondary)
            if let onEmptyTap {
                Button(action: onEmptyTap) {
                    Text("Start")
                }
                .buttonStyle(.harvestGlass(.primary))
            }
        }
        .frame(maxWidth: .infinity, minHeight: 240)
    }

    private var chart: some View {
        GeometryReader { geo in
            let size = min(geo.size.width, geo.size.height)
            let center = CGPoint(x: geo.size.width / 2, y: size / 2)
            let radius = (size / 2) - 32

            Canvas { context, _ in
                drawGrid(context: context, center: center, radius: radius)
                drawAxisLabels(context: context, center: center, radius: radius)
                if let secondary, !secondary.isZero {
                    drawPolygon(
                        context: context,
                        center: center,
                        radius: radius,
                        scores: secondary,
                        stroke: secondaryColor,
                        fill: secondaryColor.opacity(0.3)
                    )
                }
                if !primary.isZero {
                    drawPolygon(
                        context: context,
                        center: center,
                        radius: radius,
                        scores: primary,
                        stroke: primaryColor,
                        fill: primaryColor.opacity(0.3)
                    )
                }
                // Drawn last so the tier numbers stay legible over the polygons.
                drawRingNumbers(context: context, center: center, radius: radius)
            }
        }
        .frame(height: 280)
    }

    @ViewBuilder
    private var legend: some View {
        if let secondary, !secondary.isZero, let secondaryLabel {
            HStack(spacing: HarvestTheme.Spacing.lg) {
                legendDot(color: primaryColor, label: primaryLabel)
                legendDot(color: secondaryColor, label: secondaryLabel)
            }
            .frame(maxWidth: .infinity)
        } else {
            HStack(spacing: HarvestTheme.Spacing.lg) {
                legendDot(color: primaryColor, label: primaryLabel)
            }
            .frame(maxWidth: .infinity)
        }
    }

    private func legendDot(color: Color, label: String) -> some View {
        HStack(spacing: 6) {
            Circle().fill(color).frame(width: 10, height: 10)
            Text(label)
                .font(HarvestTheme.Typography.caption)
                .foregroundStyle(HarvestTheme.Colors.textSecondary)
        }
    }

    // MARK: - Geometry

    private func axisPoint(center: CGPoint, radius: Double, index: Int, magnitude: Double) -> CGPoint {
        let angle = (2 * .pi * Double(index) / Double(axes.count)) - .pi / 2
        // `magnitude` is already a radius fraction in [0, 1] (a tier ring position); clamp defensively.
        let clamped = min(max(magnitude, 0), 1)
        let r = radius * clamped
        return CGPoint(
            x: center.x + CGFloat(r * cos(angle)),
            y: center.y + CGFloat(r * sin(angle))
        )
    }

    private func drawGrid(context: GraphicsContext, center: CGPoint, radius: Double) {
        let gridColor = HarvestTheme.Colors.textSecondary.opacity(0.25)
        // Four rings, one per ValuesTier (1st → outer).
        let rings = [0.25, 0.5, 0.75, 1.0]
        for ring in rings {
            var path = Path()
            for i in 0..<axes.count {
                let p = axisPoint(center: center, radius: radius, index: i, magnitude: ring)
                if i == 0 { path.move(to: p) } else { path.addLine(to: p) }
            }
            path.closeSubpath()
            context.stroke(path, with: .color(gridColor), lineWidth: 0.5)
        }
        for i in 0..<axes.count {
            var path = Path()
            path.move(to: center)
            path.addLine(to: axisPoint(center: center, radius: radius, index: i, magnitude: 1.0))
            context.stroke(path, with: .color(gridColor), lineWidth: 0.5)
        }
    }

    private func drawRingNumbers(context: GraphicsContext, center: CGPoint, radius: Double) {
        // Tier numbers (1…4) stacked up the centre line, just left of the top spoke.
        for tier in ValuesTier.allCases {
            let y = center.y - radius * tier.radiusFraction
            let point = CGPoint(x: center.x - 12, y: y)
            let text = Text("\(tier.rawValue)")
                .font(HarvestTheme.Typography.caption)
                .foregroundStyle(HarvestTheme.Colors.textTertiary)
            context.draw(text, at: point, anchor: .center)
        }
    }

    private func drawAxisLabels(context: GraphicsContext, center: CGPoint, radius: Double) {
        for (i, axis) in axes.enumerated() {
            let labelPoint = axisPoint(center: center, radius: radius + 22, index: i, magnitude: 1.0)
            let text = Text(axis.displayName)
                .font(HarvestTheme.Typography.caption)
                .foregroundStyle(HarvestTheme.Colors.textSecondary)
            context.draw(text, at: labelPoint, anchor: .center)
        }
    }

    private func drawPolygon(
        context: GraphicsContext,
        center: CGPoint,
        radius: Double,
        scores: AxisScores,
        stroke: Color,
        fill: Color
    ) {
        var path = Path()
        for (i, axis) in axes.enumerated() {
            // Translate the raw category score into its visual tier ring position.
            let tierMagnitude = ValuesTier(rawScore: scores.value(for: axis)).radiusFraction
            let p = axisPoint(
                center: center,
                radius: radius,
                index: i,
                magnitude: tierMagnitude
            )
            if i == 0 { path.move(to: p) } else { path.addLine(to: p) }
        }
        path.closeSubpath()
        context.fill(path, with: .color(fill))
        context.stroke(path, with: .color(stroke), lineWidth: 1.5)
    }
}

#Preview("Radar — connection is a core value") {
    // Raw scores (0–28). connection 22 → Core Value (outer ring), the rest land lower.
    var s = AxisScores()
    s.emotionalIntelligence = 8; s.stability = 3; s.integrity = 14
    s.connection = 22; s.growth = 12
    return ValuesRadarCard(primary: s, primaryLabel: "I Need")
        .padding()
        .background(HarvestTheme.Colors.background)
}

#Preview("Radar — balanced (raw scores)") {
    var s = AxisScores()
    s.emotionalIntelligence = 12; s.stability = 12; s.integrity = 12
    s.connection = 12; s.growth = 12
    return ValuesRadarCard(primary: s, primaryLabel: "I Bring")
        .padding()
        .background(HarvestTheme.Colors.background)
}

#Preview("Radar — empty with action") {
    ValuesRadarCard(
        primary: AxisScores(),
        primaryLabel: "I Need",
        onEmptyTap: { print("start tapped") }
    )
    .padding()
    .background(HarvestTheme.Colors.background)
}

File: android/app/src/main/java/com/harvestglass/harvest/ui/components/ValuesRadarCard.kt
package com.harvestglass.harvest.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.background
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.BubbleChart
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.ExperimentalTextApi
import androidx.compose.ui.text.TextMeasurer
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.harvestglass.harvest.data.model.AxisScores
import com.harvestglass.harvest.data.model.ValueAxis
import com.harvestglass.harvest.data.model.ValuesTier
import com.harvestglass.harvest.ui.theme.HarvestButtonKind
import com.harvestglass.harvest.ui.theme.HarvestButton
import com.harvestglass.harvest.ui.theme.HarvestTheme
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/** Axis order as drawn, matching ValuesRadarCard.swift. */
val RADAR_AXES = listOf(
    ValueAxis.EMOTIONAL_INTELLIGENCE,
    ValueAxis.STABILITY,
    ValueAxis.INTEGRITY,
    ValueAxis.CONNECTION,
    ValueAxis.GROWTH
)

/**
 * A vertex on the radar.
 *
 * [magnitude] is already a radius fraction in [0, 1] (a tier ring position),
 * and is clamped defensively — the same guard the Swift version carries.
 */
internal fun radarAxisPoint(
    center: Offset,
    radius: Float,
    index: Int,
    axisCount: Int,
    magnitude: Float
): Offset {
    val angle = (2.0 * PI * index / axisCount) - PI / 2.0
    val clamped = magnitude.coerceIn(0f, 1f)
    val r = radius * clamped
    return Offset(
        x = center.x + (r * cos(angle)).toFloat(),
        y = center.y + (r * sin(angle)).toFloat()
    )
}

/**
 * Port of Harvest/Views/Components/ValuesRadarCard.swift.
 *
 * The chart plots visual TIERS, not raw scores: each axis's raw score (0–28)
 * goes through `ValuesTier.fromRawScore(...).radiusFraction` first, so the
 * chart communicates the *shape* of a values profile rather than points.
 */
@OptIn(ExperimentalTextApi::class)
@Composable
fun ValuesRadarCard(
    primary: AxisScores,
    primaryLabel: String,
    modifier: Modifier = Modifier,
    title: String = "Your Values Map",
    subtitle: String? = null,
    primaryColor: Color = HarvestTheme.Colors.primary,
    secondary: AxisScores? = null,
    secondaryLabel: String? = null,
    secondaryColor: Color = HarvestTheme.Colors.accent,
    onEmptyTap: (() -> Unit)? = null
) {
    val measurer = rememberTextMeasurer()
    val isEmpty = primary.isZero && (secondary?.isZero ?: true)

    GlassCard(modifier = modifier.fillMaxWidth()) {
        Column(verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.md)) {
            Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text(
                    text = title,
                    style = HarvestTheme.Typography.h4,
                    color = HarvestTheme.Colors.textPrimary
                )
                if (subtitle != null) {
                    Text(
                        text = subtitle,
                        style = HarvestTheme.Typography.bodySmall,
                        color = HarvestTheme.Colors.textSecondary
                    )
                }
            }

            if (isEmpty) {
                EmptyState(onEmptyTap)
            } else {
                RadarChart(
                    primary = primary,
                    primaryColor = primaryColor,
                    secondary = secondary,
                    secondaryColor = secondaryColor,
                    measurer = measurer
                )
                Legend(
                    primaryColor = primaryColor,
                    primaryLabel = primaryLabel,
                    secondary = secondary,
                    secondaryColor = secondaryColor,
                    secondaryLabel = secondaryLabel
                )
            }
        }
    }
}

@Composable
private fun EmptyState(onEmptyTap: (() -> Unit)?) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(
            HarvestTheme.Spacing.sm,
            Alignment.CenterVertically
        ),
        modifier = Modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = 240.dp)
    ) {
        Icon(
            imageVector = Icons.Filled.BubbleChart,
            contentDescription = null,
            tint = HarvestTheme.Colors.accent,
            modifier = Modifier.size(32.dp)
        )
        Text(
            text = "Answer a few questions to map your values.",
            style = HarvestTheme.Typography.bodyRegular,
            color = HarvestTheme.Colors.textSecondary,
            textAlign = TextAlign.Center
        )
        if (onEmptyTap != null) {
            HarvestButton(text = "Start", kind = HarvestButtonKind.PRIMARY, onClick = onEmptyTap)
        }
    }
}

@OptIn(ExperimentalTextApi::class)
@Composable
private fun RadarChart(
    primary: AxisScores,
    primaryColor: Color,
    secondary: AxisScores?,
    secondaryColor: Color,
    measurer: TextMeasurer
) {
    val gridColor = HarvestTheme.Colors.textSecondary.copy(alpha = 0.25f)
    val labelStyle = HarvestTheme.Typography.caption.copy(
        color = HarvestTheme.Colors.textSecondary
    )
    val ringNumberStyle = HarvestTheme.Typography.caption.copy(
        color = HarvestTheme.Colors.textTertiary
    )

    Canvas(
        Modifier
            .fillMaxWidth()
            .height(280.dp)
    ) {
        val side = minOf(size.width, size.height)
        val center = Offset(size.width / 2f, side / 2f)
        val radius = (side / 2f) - 32.dp.toPx()

        drawGrid(center, radius, gridColor)
        drawAxisLabels(center, radius, measurer, labelStyle)

        if (secondary != null && !secondary.isZero) {
            drawScorePolygon(center, radius, secondary, secondaryColor)
        }
        if (!primary.isZero) {
            drawScorePolygon(center, radius, primary, primaryColor)
        }

        // Drawn last so the tier numbers stay legible over the polygons.
        drawRingNumbers(center, radius, measurer, ringNumberStyle)
    }
}

private fun DrawScope.drawGrid(center: Offset, radius: Float, gridColor: Color) {
    // Four rings, one per ValuesTier (1st → outer).
    listOf(0.25f, 0.5f, 0.75f, 1f).forEach { ring ->
        val path = Path()
        RADAR_AXES.indices.forEach { i ->
            val p = radarAxisPoint(center, radius, i, RADAR_AXES.size, ring)
            if (i == 0) path.moveTo(p.x, p.y) else path.lineTo(p.x, p.y)
        }
        path.close()
        drawPath(path, gridColor, style = Stroke(width = 0.5.dp.toPx()))
    }

    RADAR_AXES.indices.forEach { i ->
        drawLine(
            color = gridColor,
            start = center,
            end = radarAxisPoint(center, radius, i, RADAR_AXES.size, 1f),
            strokeWidth = 0.5.dp.toPx()
        )
    }
}

private fun DrawScope.drawScorePolygon(
    center: Offset,
    radius: Float,
    scores: AxisScores,
    color: Color
) {
    val path = Path()
    RADAR_AXES.forEachIndexed { i, axis ->
        // Translate the raw category score into its visual tier ring position.
        val magnitude = ValuesTier.fromRawScore(scores.value(axis)).radiusFraction.toFloat()
        val p = radarAxisPoint(center, radius, i, RADAR_AXES.size, magnitude)
        if (i == 0) path.moveTo(p.x, p.y) else path.lineTo(p.x, p.y)
    }
    path.close()
    drawPath(path, color.copy(alpha = 0.3f))
    drawPath(path, color, style = Stroke(width = 1.5.dp.toPx()))
}

@OptIn(ExperimentalTextApi::class)
private fun DrawScope.drawAxisLabels(
    center: Offset,
    radius: Float,
    measurer: TextMeasurer,
    style: TextStyle
) {
    RADAR_AXES.forEachIndexed { i, axis ->
        val at = radarAxisPoint(center, radius + 22.dp.toPx(), i, RADAR_AXES.size, 1f)
        val laid = measurer.measure(axis.displayName, style)
        // Compose draws from the top-left; Swift anchors at the centre.
        drawText(
            textLayoutResult = laid,
            topLeft = Offset(at.x - laid.size.width / 2f, at.y - laid.size.height / 2f)
        )
    }
}

@OptIn(ExperimentalTextApi::class)
private fun DrawScope.drawRingNumbers(
    center: Offset,
    radius: Float,
    measurer: TextMeasurer,
    style: TextStyle
) {
    // Tier numbers (1…4) stacked up the centre line, just left of the top spoke.
    ValuesTier.entries.forEach { tier ->
        val y = center.y - radius * tier.radiusFraction.toFloat()
        val laid = measurer.measure(tier.level.toString(), style)
        drawText(
            textLayoutResult = laid,
            topLeft = Offset(
                center.x - 12.dp.toPx() - laid.size.width / 2f,
                y - laid.size.height / 2f
            )
        )
    }
}

@Composable
private fun Legend(
    primaryColor: Color,
    primaryLabel: String,
    secondary: AxisScores?,
    secondaryColor: Color,
    secondaryLabel: String?
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(
            HarvestTheme.Spacing.lg,
            Alignment.CenterHorizontally
        ),
        modifier = Modifier.fillMaxWidth()
    ) {
        LegendDot(primaryColor, primaryLabel)
        if (secondary != null && !secondary.isZero && secondaryLabel != null) {
            LegendDot(secondaryColor, secondaryLabel)
        }
    }
}

@Composable
private fun LegendDot(color: Color, label: String) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(Modifier.size(10.dp).background(color, CircleShape))
        Text(
            text = label,
            style = HarvestTheme.Typography.caption,
            color = HarvestTheme.Colors.textSecondary
        )
    }
}
